File: 528_pick_index/pick_index.go
/*
 * @Date: 2021-08-30 13:03:45
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-30 13:09:08
 */

package main

import (
	"math/rand"
	"sort"
)

type Solution struct {
	pre []int
}

func Constructor(w []int) Solution {
	for i := 1; i < len(w); i++ {
		w[i] += w[i-1]
	}
	return Solution{w}
}

func (s *Solution) PickIndex() int {
	x := rand.Intn(s.pre[len(s.pre)-1]) + 1
	return sort.SearchInts(s.pre, x)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		s := Constructor([]int{1})
		assert(s.PickIndex() == 0)
	}
	{
		s := Constructor([]int{1, 3})
		assert(s.PickIndex() == 1)
	}
}

File: cpp/528_pick_index/pick_index.cpp
/*
 * @Date: 2021-08-30 12:49:04
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-30 13:10:20
 * @FilePath: /algorithm/528_pick_index/pick_index.cpp
 * @Description: file content
 */

#include <cassert>
#include <random>
#include <vector>

using namespace std;

class Solution {
 private:
  mt19937 gen;
  uniform_int_distribution<int> dis;
  vector<int> pre;

 public:
  Solution(vector<int>& w)
      : gen(random_device{}()), dis(1, accumulate(w.begin(), w.end(), 0)) {
    partial_sum(w.begin(), w.end(), back_inserter(pre));
  }

  int pickIndex() {
    int x = dis(gen);
    return lower_bound(pre.begin(), pre.end(), x) - pre.begin();
  }
};

int main() {
  {
    vector<int> nums{1};
    Solution s{nums};
    assert(s.pickIndex() == 0);
  }
  {
    vector<int> nums{1, 3};
    Solution s{nums};
    assert(s.pickIndex() == 1);
  }
}